core/src/View/MDSpan/Kokkos_MDSpan_Accessor.hpp

```diff
@@ -212,6 +212,334 @@ struct AtomicAccessorRelaxed {
   }
 };
 
+//=====================================================================
+//============= Reference Counted Accessor and DataHandle =============
+//=====================================================================
+
+template <class ElementType, class MemorySpace>
+class ReferenceCountedDataHandle {
+ public:
+  using value_type   = ElementType;
+  using pointer      = value_type*;
+  using reference    = value_type&;
+  using memory_space = MemorySpace;
+
+  KOKKOS_DEFAULTED_FUNCTION
+  ReferenceCountedDataHandle() = default;
+  explicit ReferenceCountedDataHandle(SharedAllocationRecord<void, void>* rec) {
+    m_tracker.assign_allocated_record_to_uninitialized(rec);
+    m_handle = static_cast<pointer>(get_record()->data());
+  }
+
+  KOKKOS_FUNCTION
+  ReferenceCountedDataHandle(const SharedAllocationTracker& tracker,
+                             pointer data_handle)
+      : m_tracker(tracker), m_handle(data_handle) {}
+
+  // unmanaged ctor
+  template <class OtherElementType,
+            class = std::enable_if_t<std::is_convertible_v<
+                OtherElementType (*)[], value_type (*)[]>>>
+  KOKKOS_FUNCTION ReferenceCountedDataHandle(OtherElementType* ptr)
+      : m_tracker(), m_handle(ptr) {}
+
+  // subview ctor
+  template <class OtherElementType,
+            class = std::enable_if_t<std::is_convertible_v<
+                OtherElementType (*)[], value_type (*)[]>>>
+  KOKKOS_FUNCTION ReferenceCountedDataHandle(
+      const ReferenceCountedDataHandle& other, OtherElementType* ptr)
+      : m_tracker(other.m_tracker), m_handle(ptr) {}
+
+  // converting ctor
+  template <class OtherElementType,
+            class = std::enable_if_t<std::is_convertible_v<
+                OtherElementType (*)[], value_type (*)[]>>>
+  KOKKOS_FUNCTION ReferenceCountedDataHandle(
+      const ReferenceCountedDataHandle<OtherElementType, memory_space>& other)
+      : m_tracker(other.m_tracker), m_handle(other.m_handle) {}
+
+  KOKKOS_DEFAULTED_FUNCTION
+  ReferenceCountedDataHandle(const ReferenceCountedDataHandle&) = default;
+  KOKKOS_DEFAULTED_FUNCTION
+  ReferenceCountedDataHandle(ReferenceCountedDataHandle&&) noexcept = default;
+  KOKKOS_DEFAULTED_FUNCTION
+  ReferenceCountedDataHandle& operator=(const ReferenceCountedDataHandle&) =
+      default;
+  KOKKOS_DEFAULTED_FUNCTION
+  ReferenceCountedDataHandle& operator=(ReferenceCountedDataHandle&&) = default;
+
+  KOKKOS_FUNCTION
+  pointer get() const noexcept { return m_handle; }
+  KOKKOS_FUNCTION
+  explicit operator pointer() const noexcept { return m_handle; }
+
+  bool has_record() const { return m_tracker.has_record(); }
+  auto* get_record() const { return m_tracker.get_record<memory_space>(); }
+  int use_count() const noexcept { return m_tracker.use_count(); }
+
+  std::string get_label() const { return m_tracker.get_label<memory_space>(); }
```

Comment:
What about this ?
```suggestion
  const std::string& get_label() const { return m_tracker.get_label<memory_space>(); }
```

Reply:
OK, so the debate might probably move to why `Kokkos` makes new strings each time we ask for the label ? I guess that some allocation shared records store a `char*` to be used on device, so why not returning a string view ?

Reply:
Noting that view labels are usually quite long.

Reply:
The string might actually come from device storage copied over into a temporary. Generally, labels are a debug functionality so we were never super concerned about making them fast.

Reply:
OK, never mind.